StripePaymentSheet/StripePaymentSheet/Source/Internal/ECE/ECEViewController.swift

```diff
@@ -0,0 +1,533 @@
+//
+//  ECEViewController.swift
+//  StripePaymentSheet
+//
+
+@_spi(STP) import StripeCore
+@_spi(STP) import StripeUICore
+import UIKit
+import WebKit
+
+// Protocol for handling Express Checkout Element events
+@available(iOS 16.0, *)
+protocol ExpressCheckoutWebviewDelegate: AnyObject {
+    func amountForECEView(_ eceView: ECEViewController) -> Int
+    func eceView(_ eceView: ECEViewController, didReceiveShippingAddressChange shippingAddress: [String: Any]) async throws -> [String: Any]
+    func eceView(_ eceView: ECEViewController, didReceiveShippingRateChange shippingRate: [String: Any]) async throws -> [String: Any]
+    func eceView(_ eceView: ECEViewController, didReceiveECEClick event: [String: Any]) async throws -> [String: Any]
+    func eceView(_ eceView: ECEViewController, didReceiveECEConfirmation paymentDetails: [String: Any]) async throws -> [String: Any]
+}
+
+// Custom errors for Express Checkout operations
+enum ExpressCheckoutError: LocalizedError {
+    case invalidShippingRate(rateId: String)
+    case missingRequiredField(field: String)
+
+    var errorDescription: String? {
+        switch self {
+        case .invalidShippingRate(let rateId):
+            return "Invalid shipping rate: \(rateId)"
+        case .missingRequiredField(let field):
+            return "Missing required field: \(field)"
+        }
+    }
+}
+
+@available(iOS 16.0, *)
+class ECEViewController: UIViewController {
+    let apiClient: STPAPIClient
+
+    init(apiClient: STPAPIClient) {
+        self.apiClient = apiClient
+        super.init(nibName: nil, bundle: nil)
+    }
+
+    required init?(coder: NSCoder) {
+        fatalError("init(coder:) has not been implemented")
+    }
+
+    private var webView: WKWebView!
+    private var popupWebView: WKWebView?
+
+    // Delegate for Express Checkout Element events
+    weak var expressCheckoutWebviewDelegate: ExpressCheckoutWebviewDelegate?
+
+    // We normally hide Express Checkout Element in WKWebViews (as many don't handle popups correctly)
+    // Fake Safari's UA to disable that behavior
+    static let FakeSafariUserAgent = "Mozilla/5.0 (iPhone; CPU iPhone OS 18_1 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/18.1.1 Mobile/15E148 Safari/604.1"
```

Comment:
are there plans to fix this pre-launch?

Reply:
It might be better to add a specific tag to the user agent instead of impersonating Safari, let's coordinate with ECE folks